transport.go

```diff
@@ -17,75 +17,118 @@ const LegacyTransportID TransportID = "graphsync"
-	// return values are:
-	// - nil = proceed with sending data
-	// - error = cancel this request
-	// - err == ErrPause - pause this request
-	OnDataReceived(chid ChannelID, link ipld.Link, size uint64, index int64, unique bool) error
-
-	// OnDataQueued is called when data is queued for sending for the given channel ID
-	// return values are:
-	// - nil = proceed with sending data
-	// - error = cancel this request
-	// - err == ErrPause - pause this request
-	OnDataQueued(chid ChannelID, link ipld.Link, size uint64, index int64, unique bool) (Message, error)
-
-	// OnDataSent is called when we send data for the given channel ID
-	OnDataSent(chid ChannelID, link ipld.Link, size uint64, index int64, unique bool) error
-
-	// OnTransferQueued is called when a new data transfer request is queued in the transport layer.
-	OnTransferQueued(chid ChannelID)
-
-	// OnRequestReceived is called when we receive a new request to send data
-	// for the given channel ID
-	// return values are:
-	// message = data transfer message along with reply
-	// err = error
-	// - nil = proceed with sending data
-	// - error = cancel this request
-	// - err == ErrPause - pause this request (only for new requests)
-	// - err == ErrResume - resume this request (only for update requests)
+	// OnRequestReceived occurs when we receive a  request for the given channel ID
+	// return values are a message to send an error if the transport should be closed
 	OnRequestReceived(chid ChannelID, msg Request) (Response, error)
-	// OnChannelCompleted is called when we finish transferring data for the given channel ID
-	// Error returns are logged but otherwise have no effect
-	OnChannelCompleted(chid ChannelID, err error) error
 
-	// OnRequestCancelled is called when a request we opened (with the given channel Id) to
-	// receive data is cancelled by us.
-	// Error returns are logged but otherwise have no effect
-	OnRequestCancelled(chid ChannelID, err error) error
-
-	// OnRequestDisconnected is called when a network error occurs trying to send a request
-	OnRequestDisconnected(chid ChannelID, err error) error
-
-	// OnSendDataError is called when a network error occurs sending data
-	// at the transport layer
-	OnSendDataError(chid ChannelID, err error) error
+	// OnRequestReceived occurs when we receive a response to a request
+	OnResponseReceived(chid ChannelID, msg Response) error
 
-	// OnReceiveDataError is called when a network error occurs receiving data
-	// at the transport layer
-	OnReceiveDataError(chid ChannelID, err error) error
+	// OnTransportEvent is dispatched when an event occurs on the transport
+	// It MAY be dispatched asynchronously by the transport to the time the
+	// event occurs
+	// However, the other handler functions may ONLY be called on the same channel
+	// after all events are dispatched. In other words, the transport MUST allow
+	// the handler to process all events before calling the other functions which
+	// have a synchronous return
```

Comment:
From your latest comments in the thread, I think this comment is getting at the point you made:

> Where I would like to get to is no synchronous hooks. Just one OnTransportEvent function.

Which points to the deficiency in the current half-way state? I take it that you've chewed off the doable events into `OnTransportEvent()`, with the remaining ones TODO.

So we still have:

* OnRequestReceived
* OnRequestReceived
* OnContextAugment

And you've experimented with splitting `OnRequestReceived` into `Update`, `Pause` and `Voucher` forms.

The context augment one being synchronous makes sense, the other ones do stand out a little now so I see the awkwardness of not going the full way on bundling everything into the single event _channel_. But, I think it's reasonable to stop where you have just to get this over the line.

However, it would be good to add some more inline comments about this state as a reminder for later or a hint for the next person picking this up about (a) the fact of the half-way state and (b) the ideal of moving toward fully async eventing.

Reply:
absolutely. yes, the hard part about the others being asynchronous is that graphsync has synchronous hooks. so we'll need a bit of channel magic. messing with all this at once was just creating a back and forth with no progress being made.